### DISCUSSION TITLE
Polygon edges showing on composite image

### DISCUSSION THREAD

**Discussion by eastwestsoftware** [ImageMagick/ImageMagick#2486] [Help] [answered]
I am facing a problem with edges showing on the result of a composite command.

**Version information:**
Version: ImageMagick 6.9.7-4 Q16 x86_64 20170114
Features: Cipher DPC Modules OpenMP 
Delegates (built-in): bzlib djvu fftw fontconfig freetype jbig jng jpeg lcms lqr ltdl lzma openexr pangocairo png tiff wmf x xml zlib
OS: Ubuntu 18.04.4 LTS

**This is the code, please see explanation below:**

```
#!/bin/bash

left=200
right=800
top=200
bottom=800
polygonmargin=50

polygon="
    $((left + polygonmargin)),$((top))
    $((right - polygonmargin)),$((top))
    $((right)),$((top + polygonmargin))
    $((right)),$((bottom - polygonmargin))
    $((right - polygonmargin)),$((bottom))
    $((left + polygonmargin)),$((bottom))
    $((left)),$((bottom - polygonmargin))
    $((left)),$((top + polygonmargin))
    $((left + polygonmargin)),$((top))
"
convert -size 1000x1000 xc:none \
    \( -fill black -draw "polygon $polygon" \) \
    \( -fill black -draw "rectangle 0,0 10,10" \) \
mask_test.png

composite -compose dst-out mask_test.png rect/original.jpg -matte outer_test.png
composite -compose dst-in mask_test.png rect/original.jpg -matte inner_test.png

convert inner_test.png -fill none -draw "matte 0,0 replace" inner_transparent_test.png
convert outer_test.png -fill none -opaque "#FFFFFF" outer_transparent_test.png

convert inner_transparent_test.png outer_transparent_test.png -composite result_test.jpg
```

final result:

![result_test](https://user-images.githubusercontent.com/26831029/91706236-64eece00-eb7e-11ea-87b3-ad796591dd9f.jpg)

Is there a way to avoid the edges being displayed on the final image?

**----- Explanation -----**

Here are the steps:

This is the source image (original.jpg):

![original](https://user-images.githubusercontent.com/26831029/91704456-ef81fe00-eb7b-11ea-95fa-6f6ad24d64e8.jpg)

I am creating a polygon mask with this code:

```
convert -size 1000x1000 xc:none \
    \( -fill black -draw "polygon $polygon" \) \
    \( -fill black -draw "rectangle 0,0 10,10" \) \
mask_test.png
```

(the rectangle on the top is used for defining the color for the matte command below)

![mask_test](https://user-images.githubusercontent.com/26831029/91704346-c7929a80-eb7b-11ea-836c-b6fd50f7eed7.png)

Then I use the mask to split an image into 2 parts, inner and outer

`composite -compose dst-out mask_test.png original.jpg -matte outer_test.png`
`composite -compose dst-in mask_test.png original.jpg -matte inner_test.png`

![outer_test](https://user-images.githubusercontent.com/26831029/91704970-98305d80-eb7c-11ea-9d51-ef0ddaadc1ab.png)
(above here is a white image with a transparent box in the middle - just not very clear in github)
![inner_test](https://user-images.githubusercontent.com/26831029/91704978-9a92b780-eb7c-11ea-85b3-b742ceb71d75.png)
(above here is a transparent image with a white box and the black circle)

After that, I apply 2 different fill techniques:

```
convert inner_test.png -fill none -draw "matte 0,0 replace" inner_transparent_test.png
convert outer_test.png -fill none -opaque "#FFFFFF" outer_transparent_test.png
```

Results:

![inner_transparent_test](https://user-images.githubusercontent.com/26831029/91705646-8dc29380-eb7d-11ea-913d-8a032515db5e.png)
(above here is the black circle from the original image on a transparent background)
![outer_transparent_test](https://user-images.githubusercontent.com/26831029/91705660-91561a80-eb7d-11ea-8744-46e69a408ef7.png)
(above here is a transparent box on a black background)
here the problem shows already: white edges on the polygon

finally, I composite the 2 images to the final result:

`convert inner_transparent_test.png outer_transparent_test.png -composite result_test.jpg`

![result_test](https://user-images.githubusercontent.com/26831029/91705904-eb56e000-eb7d-11ea-970e-314a26881425.jpg)

**Comment by fmw42** [comment, marked as answer]
I believe your issue is due to using antialiased drawings, then trying to fill them. The fill will only fill the pure colors and not the antialiased edges. Try drawing with +antialias in your commands. Since you did not provide the polygon coordinates, we cannot test your code properly. Also I would suggest you use convert ... -composite rather than composite. You should be able to make that one convert command line.  Also please identify your exact IM version and platform.

**Comment by eastwestsoftware** [comment]
Thank you, your advice to add +antialias to the draw commands worked.
